src/core/ext/xds/xds_http_filters.cc

```diff
@@ -0,0 +1,93 @@
+//
+// Copyright 2021 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/ext/xds/xds_http_filters.h"
+
+#include "envoy/extensions/filters/http/router/v3/router.upbdefs.h"
+
+namespace grpc_core {
+
+const char* kXdsHttpRouterFilterConfigName =
+    "extensions.filters.http.router.v3.Router";
+
+namespace {
+
+class XdsHttpRouterFilter : public XdsHttpFilterImpl {
+ public:
+  absl::string_view config_proto_type_name() const override {
+    return kXdsHttpRouterFilterConfigName;
+  }
+
+  void PopulateSymtab(upb_symtab* symtab) const override {
+    envoy_extensions_filters_http_router_v3_Router_getmsgdef(symtab);
+  }
+
+  absl::StatusOr<FilterConfig> GenerateFilterConfig(
+      upb_strview serialized_xds_config, upb_arena* arena) const override {
+    return FilterConfig{kXdsHttpRouterFilterConfigName, Json()};
+  }
+
+  absl::StatusOr<FilterConfig> GenerateFilterConfigOverride(
+      upb_strview serialized_xds_config, upb_arena* arena) const override {
+    return FilterConfig{kXdsHttpRouterFilterConfigName, Json()};
+  }
+
+  // No-op -- this filter is special-cased by the xds resolver.
+  const grpc_channel_filter* channel_filter() const override { return nullptr; }
+
+  // No-op -- this filter is special-cased by the xds resolver.
+  absl::StatusOr<std::string> GenerateServiceConfig(
+      const FilterConfig& hcm_filter_config,
+      const FilterConfig* filter_config_override) const override {
```

Comment:
May not apply to this function particular, but should we check the type of the `FilterConfig` and the config proto type name? Ideally, we could use some mechanism to prevent passing wrong `FilterConfig` down in build time.

Reply:
As I mentioned above, some filters use a different proto message type for their top-level config and for their override config, so this isn't something we can check for them -- the individual filter implementation will have to do that.

That reminds me, I had meant to make the registry use a shared_ptr, so that the same filter implementation can be registered under two different config proto names.  I've gone ahead and made that change.